src/main/java/org/opensearch/knn/index/util/Faiss.java

```diff
@@ -0,0 +1,327 @@
+                            throw new IllegalStateException(
+                                String.format("%s  is not a valid parameter. This is a bug.", METHOD_PARAMETER_NLIST)
+                            );
+                        }
+
+                        nlistObject = nlistParameter.getDefaultValue();
+                    }
+
+                    if (!(nlistObject instanceof Integer)) {
+                        throw new IllegalStateException(String.format("%s must be an integer.", METHOD_PARAMETER_NLIST));
+                    }
+
+                    int centroids = (Integer) nlistObject;
+                    return ((4L * centroids * dimension) / BYTES_PER_KILOBYTES) + 1;
+                })
+                .build()
+        ).addSpaces(SpaceType.L2, SpaceType.INNER_PRODUCT).build()
+    );
+
+    final static Faiss INSTANCE = new Faiss(METHODS, SCORE_TRANSLATIONS, CURRENT_VERSION, KNNConstants.FAISS_EXTENSION);
+
+    /**
+     * Constructor for Faiss
+     *
+     * @param methods                   map of methods the native library supports
+     * @param scoreTranslation          Map of translation of space type to scores returned by the library
+     * @param currentVersion            String representation of current version of the library
+     * @param extension                 String representing the extension that library files should use
+     */
+    private Faiss(
+        Map<String, KNNMethod> methods,
+        Map<SpaceType, Function<Float, Float>> scoreTranslation,
+        String currentVersion,
+        String extension
+    ) {
+        super(methods, scoreTranslation, currentVersion, extension);
+    }
+
+    /**
+     * MethodAsMap builder is used to create the map that will be passed to the jni to create the faiss index.
+     * Faiss's index factory takes an "index description" that it uses to build the index. In this description,
+     * some parameters of the index can be configured; others need to be manually set. MethodMap builder creates
+     * the index description from a set of parameters and removes them from the map. On build, it sets the index
+     * description in the map and returns the processed map
+     */
+    @AllArgsConstructor
+    static class MethodAsMapBuilder {
+        String indexDescription;
+        MethodComponent methodComponent;
+        Map<String, Object> methodAsMap;
+
+        /**
+         * Add a parameter that will be used in the index description for the given method component
+         *
+         * @param parameterName name of the parameter
+         * @param prefix to append to the index description before the parameter
+         * @param suffix to append to the index description after the parameter
+         * @return this builder
+         */
+        @SuppressWarnings("unchecked")
+        MethodAsMapBuilder addParameter(String parameterName, String prefix, String suffix) {
+            indexDescription += prefix;
```

Comment:
nit: maybe we can refactor in one of next PRs to something more efficient like StringBuilder. Seems we're accumulating multiple small strings and concatenate all of them into one final string object.

Reply:
That makes sense. Will take for future PR.